src/katgpucbf/dsim/signal.py

```diff
@@ -198,28 +200,73 @@ def __str__(self) -> str:
 
 
 @dataclass
-class CW(Signal):
-    """Continuous wave.
+class Periodic(Signal):
+    """Base class for period signals.
 
-    To make the resulting signal periodic, the frequency is adjusted during
-    sampling so that the sampled result can be looped.
+    The frequency is adjusted during sampling so that the sampled result can
+    be looped.
     """
 
     amplitude: float
     frequency: float
+    _class_name: ClassVar[str] = ""
 
     @staticmethod
-    def _sample_chunk(offset: np.ndarray, *, amplitude: np.float32, chunk_size: int, scale: float) -> np.ndarray:
-        """Compute :math:`np.cos(np.arange(offset, n + offset) * scale) * amplitude` efficiently.
+    @abstractmethod
+    def _sample_chunk(offset: np.ndarray, *, amplitude: np.float32, chunk_size: int, frequency: float) -> np.ndarray:
+        """Compute a single chunk.
 
-        The return value is single precision. `offset` must be an array with a
-        single element.
+        Parameters
+        ----------
+        offset
+            Index of the first element of the chunk (1-element array)
+        amplitude
+            Amplitude of the signal
+        chunk_size
+            Number of samples to produce
+        frequency
+            Rounded frequency, expressed as cycles per sample
         """
```

Comment:
The fact that this is passing pre-commit suggests that this question is pointless, but I was under the impression that an `@abstractmethod` needed at the very least a `...` or `pass`. Or does the docstring serve that purpose?

Reply:
Yep, a docstring is enough to keep the parser happy - it counts as a "statement"